### PR TITLE
assumeutxo: Fix -reindex before snapshot was validated

### PR DESCRIPTION
**Pull request by mzumsande**
In c711ca186f8d8a28810be0beedcb615ddcf93163 logic was introduced that `-reindex` and `-reindex-chainstate` will delete the snapshot chainstate.
This doesn't work currently, instead of deleting the snapshot chainstate the node crashes with an assert (this can be triggered by applying the added test commit on master).
Fix this, and another bug that would prevent the new active chainstate from having a mempool after `-reindex` has deleted the snapshot (also covered by the test).

### CODE REVIEW

**Review by BrandonOdiwuor — COMMENTED** [review]
Concept ACK

File: test/functional/feature_assumeutxo.py [hunk 1]
```diff
@@ -278,6 +275,18 @@ def check_tx_counts(final: bool) -> None:
 
         assert_equal(n1.getblockchaininfo()["blocks"], SNAPSHOT_BASE_HEIGHT)
 
+        self.log.info(f"Check that restarting with -reindex will delete the snapshot chainstate")
+        self.restart_node(1, extra_args=['-reindex=1', *self.extra_args[1]])
```
Comment (fjahr) [inline test/functional/feature_assumeutxo.py:279]: This only covers `-reindex` but not `-reindex-chainstate`. It's easily possible when using node 2 for the test which is not pruned: https://github.com/fjahr/bitcoin/commit/38c5d80f73c7058f17ba8e89a3802ef981080864

**Review by fjahr — COMMENTED** [review]
tACK cfdef5c5d7a2d2e0bc78aa2cde534e541af61d1c

Good catch! I could reproduce the crash and that the change here fixes this.

My suggested test extension can be pulled in here if you retouch or I can add it to one of my outstanding assumeutxo PRs as a small follow-up.

**Review by BrandonOdiwuor — COMMENTED** [review]
crACK cfdef5c5d7a2d2e0bc78aa2cde534e541af61d1c

File: test/functional/feature_assumeutxo.py (hunk 1 above)
Comment (mzumsande) [reply test/functional/feature_assumeutxo.py]: Thanks! I took your suggestion (with a minor change to logging).

**Comment by mzumsande** [comment]
[cfdef5c ](https://github.com/bitcoin/bitcoin/commit/cfdef5c5d7a2d2e0bc78aa2cde534e541af61d1c)to [b7ba60f](https://github.com/bitcoin/bitcoin/commit/b7ba60f81a33db876f88b5f9af1e5025d679b5be): Included test coverage for `-reindex-chainstate` as suggested by @fjahr

**Comment by fjahr** [comment]
re-ACK b7ba60f81a33db876f88b5f9af1e5025d679b5be

**Review by BrandonOdiwuor — APPROVED** [review]
re-ACK b7ba60f81a33db876f88b5f9af1e5025d679b5be

**Comment by hernanmarino** [comment]
crACK b7ba60f81a33db876f88b5f9af1e5025d679b5be . Good fix

**Review by byaye — APPROVED** [review]
Tested ACK b7ba60f81a33db876f88b5f9af1e5025d679b5be

Test result before the fix
Screenshot 2024-04-12 at 17 34 14

Test result after the fix
Screenshot 2024-04-12 at 17 21 48

**Pull request merged**